### DISCUSSION TITLE
How to use options creator functions with new typegen types

### DISCUSSION THREAD

**Discussion by SimeonC** [statelyai/xstate#3086] [Q&A] [answered]
At our company we have a 300 line limit for files, so something I've often done is create a compose function to put the services/guards/actions creation in another file. With xstate 4.30 I can't seem to get the types for this to work anymore. Here's an example of what I do;

```tsx
export function createCustomMachine({ someFunction }): StateMachine<...> {
  // some setup code
  return createMachine({
    schema: {
      context: {} as Context,
      events: {} as Events
    },
    initial: 'init',
    states: { ... }
  }, {
    services: composeServices(someFunction)
  });
}

// in another file
export function composeServices(someFunction): Record<
  string,
  ServiceConfig<
    Context,
    Events
  > {
  ...
}
```

This results in the following typescript error occurring on `services: composeServices(someFunction)`.

```
TS2322: Type 'Record<string, ServiceConfig<Context<TFieldValues, TTabType>, Event<TFieldValues, TSubmitReturn, TTabType>>>' is not assignable to type 'MachineOptionsServices<Context<TFieldValues, TTabType>, TypegenDisabled & { missingImplementations: { actions: never; delays: never; guards: never; services: never; }; } & AllowAllEvents & { ...; }, Record<...>, Record<...> & { ...; }, Record<...>>'.
   'string' index signatures are incompatible.
     Type 'ServiceConfig<Context<TFieldValues, TTabType>, Event<TFieldValues, TSubmitReturn, TTabType>>' is not assignable to type 'AnyStateMachine | InvokeCreator<Context<TFieldValues, TTabType>, { type: "TOGGLE_SECTION"; id: string; } | { type: "SUBMIT" | "NEXT"; } | ... 6 more ... | ErrorFocusEvent<...>, any, EventObject, { ...; } | ... 7 more ... | ErrorFocusEvent<...>> | undefined'.
       Type 'string' is not assignable to type 'AnyStateMachine | InvokeCreator<Context<TFieldValues, TTabType>, { type: "TOGGLE_SECTION"; id: string; } | { type: "SUBMIT" | "NEXT"; } | ... 6 more ... | ErrorFocusEvent<...>, any, EventObject, { ...; } | ... 7 more ... | ErrorFocusEvent<...>> | undefined'.
```

**Comment by Andarist** [comment]
Could u share a full runnable repro case of the problem? That would ensure that I would comment on your exact problem.

**Reply by SimeonC** [reply]
Sure, here you go https://codesandbox.io/s/sharp-tree-7eeyi2?file=/src/index.ts (I really should do this automatically but I keep getting lazy and forgetting 😓 )

**Reply by SimeonC** [reply]
Ah, I should also mention that our linter rules have the `@typescript-eslint/explicit-module-boundary-types` turned on, so the `composeServices` function requires a return type.

**Comment by SimeonC** [comment, marked as answer]
I realised my mistake and this fixed the issue in the upgrade.

```ts
service: Record<
  string,
  ServiceConfig<
    Context,
    Events
  >
```

Should now be the following.

```ts
service: Record<
  string,
  AnyStateMachine | InvokeCreator<
    Context,
    Events
  >
```

**Reply by Andarist** [reply]
Yeah, `ServiceConfig` also includes `string` in the union, and plain strings are not allowed at this position. So if this worked in the past - it was because XState types were not accurate here.

However, note that the used structure isn't super safe and won't allow you to utilize the full power of typegen. If you plan on using the typegen (which I encourage you to try out very much) then you might reconsider how do you write this sort of stuff. The problem here is that the correct type for the "implementations"/options is not typeable by hand (well, it is - but you'd have to update it manually after changing your machine definition so I wouldn't consider it as a viable option).

We can unpack this type from the machine type using `MachineOptionsFrom`, like here: https://codesandbox.io/s/intelligent-aryabhata-2sbhgz?file=/src/index.ts . This is nice because it should provide you strict and auto-synced types when you start using the typegen.

At the moment there is a caveat to this solution though - `withConfig` requires you to provide **all** the required implementations. So you can't partially provide some implementations here, while deferring providing the rest until some other call, like `useMachine` etc

**Reply by SimeonC** [reply]
Awesome, thanks for the tip about `MachineOptionsFrom` that should save a lot of my type dancing, though I will have to refactor my file structure a bit it looks like it'll be better overall.

I'm looking forward to using typegen when I can, but we don't have all our devs using VS Code so I'm waiting for a CLI. (Though the new typegen and viz tools have sorely tempted me to have another look at switching from intellij 😆 )
